File: android/src/main/java/com/shtreb/socket_io/SocketIoPlugin.java
package com.shtreb.socket_io;

import androidx.annotation.NonNull;

import java.util.Map;

import com.google.gson.Gson;

import io.flutter.embedding.engine.plugins.FlutterPlugin;
import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;
import io.flutter.plugin.common.MethodChannel.MethodCallHandler;
import io.flutter.plugin.common.MethodChannel.Result;
import io.flutter.plugin.common.PluginRegistry.Registrar;

/** SocketIoPlugin */
public class SocketIoPlugin implements FlutterPlugin, MethodCallHandler {
  public static final String LOG_TAG = SocketIoPlugin.class.getName();
  private MethodChannel channel;

  @Override
  public void onAttachedToEngine(@NonNull FlutterPluginBinding flutterPluginBinding) {
    channel = new MethodChannel(flutterPluginBinding.getBinaryMessenger(), "socket_io");
    channel.setMethodCallHandler(this);
  }

  @Override
  public void onMethodCall(@NonNull MethodCall call, @NonNull Result result) {
    String socketNameSpace = call.argument(SocketIOManager.MethodCallArgumentsName.SOCKET_NAME_SPACE);
    String socketDomain = call.argument(SocketIOManager.MethodCallArgumentsName.SOCKET_DOMAIN);
    String callback = call.argument(SocketIOManager.MethodCallArgumentsName.SOCKET_CALLBACK);

    Utils.log(LOG_TAG, "onMethodCall: " + call.method + " - domain: " + socketDomain + " - with namespace: " + socketNameSpace);

    switch (call.method) {
      case SocketIOManager.MethodCallName.SOCKET_INIT:
        String query = call.argument(SocketIOManager.MethodCallArgumentsName.SOCKET_QUERY);
        SocketIOManager.getInstance().init(channel, socketDomain, socketNameSpace, query, callback);
        break;

      case SocketIOManager.MethodCallName.SOCKET_CONNECT:
        SocketIOManager.getInstance().connect(socketDomain, socketNameSpace);
        break;

      case SocketIOManager.MethodCallName.SOCKET_DISCONNECT:
        SocketIOManager.getInstance().disconnect(socketDomain, socketNameSpace);
        break;

      case SocketIOManager.MethodCallName.SOCKET_SUBSCRIBES:
        String socketData = call.argument(SocketIOManager.MethodCallArgumentsName.SOCKET_DATA);
        Map<String, String> map = Utils.convertJsonToMap(socketData);
        Utils.log(LOG_TAG, "socketData: " + new Gson().toJson(map));
        SocketIOManager.getInstance().subscribes(socketDomain, socketNameSpace, map);
        break;

      case SocketIOManager.MethodCallName.SOCKET_UNSUBSCRIBES_ALL:
        SocketIOManager.getInstance().unSubscribesAll(socketDomain, socketNameSpace);
        break;

      case SocketIOManager.MethodCallName.SOCKET_UNSUBSCRIBES:
        String jsonData = call.argument(SocketIOManager.MethodCallArgumentsName.SOCKET_DATA);
        Map<String, String> params = Utils.convertJsonToMap(jsonData);
        SocketIOManager.getInstance().unSubscribes(socketDomain, socketNameSpace, params);
        break;

      case SocketIOManager.MethodCallName.SOCKET_SEND_MESSAGE:
        String event = call.argument(SocketIOManager.MethodCallArgumentsName.SOCKET_EVENT);
        String message = call.argument(SocketIOManager.MethodCallArgumentsName.SOCKET_MESSAGE);
        if (!Utils.isNullOrEmpty(event) && message != null) {
          SocketIOManager.getInstance().sendMessage(socketDomain, socketNameSpace, event, message, callback);
        } else {
          Utils.log(LOG_TAG, "send message with invalid params:" + "Event: " + event + " - with message: " + new Gson().toJson(message));
        }
        break;

      case SocketIOManager.MethodCallName.SOCKET_DESTROY:
        SocketIOManager.getInstance().destroySocket(socketDomain, socketNameSpace);
        break;

      case SocketIOManager.MethodCallName.SOCKET_DESTROY_ALL:
        SocketIOManager.getInstance().destroyAllSockets();
        break;

      default:
        result.notImplemented();
        break;
    }
  }

  @Override
  public void onDetachedFromEngine(@NonNull FlutterPluginBinding binding) {
    channel.setMethodCallHandler(null);
  }
}

File: ios/Classes/SocketIoPlugin.m
#import "SocketIoPlugin.h"
#import "SocketObj/SocketIOManager.h"

static FlutterMethodChannel *channel;

@implementation SocketIoPlugin

+ (void)registerWithRegistrar:(NSObject<FlutterPluginRegistrar>*)registrar {
    channel = [FlutterMethodChannel
               methodChannelWithName:@"socket_io"
               binaryMessenger:[registrar messenger]];
    SocketIoPlugin* instance = [[SocketIoPlugin alloc] init];
    [registrar addMethodCallDelegate:instance channel:channel];
}

- (void)handleMethodCall:(FlutterMethodCall*)call result:(FlutterResult)result {
    NSString *socketNameSpace = call.arguments[SOCKET_NAME_SPACE];
    NSString *socketDomain = call.arguments[SOCKET_DOMAIN];
    NSString *callback = call.arguments[SOCKET_CALLBACK];
    if ([SOCKET_INIT isEqualToString:call.method]) {
        //NSString *query = @"userId=123123&abc=xxx";
        NSString *query = call.arguments[SOCKET_QUERY];
        NSMutableDictionary *queryStringDictionary = [[NSMutableDictionary alloc] init];
        NSArray *urlComponents = [query componentsSeparatedByString:@"&"];
        for (NSString *keyValuePair in urlComponents) {
            NSArray *pairComponents = [keyValuePair componentsSeparatedByString:@"="];
            NSString *key = [[pairComponents firstObject] stringByRemovingPercentEncoding];
            NSString *value = [[pairComponents lastObject] stringByRemovingPercentEncoding];
            
            [queryStringDictionary setObject:value forKey:key];
        }
        [[SocketIOManager sharedSocketIOManager] initSocket:channel domain:socketDomain query:queryStringDictionary namspace:socketNameSpace callBackStatus:callback];
    } else if ([SOCKET_CONNECT isEqualToString:call.method]) {
        [[SocketIOManager sharedSocketIOManager] connectSocket:socketDomain namspace:socketNameSpace];
    } else if ([SOCKET_DISCONNECT isEqualToString:call.method]) {
        [[SocketIOManager sharedSocketIOManager] disconnectDomain:socketDomain namspace:socketNameSpace];
    } else if ([SOCKET_SUBSCRIBES isEqualToString:call.method]) {
        NSString *socketData = call.arguments[SOCKET_DATA];
        NSData *data = [socketData dataUsingEncoding:NSUTF8StringEncoding];
        id map = [NSJSONSerialization JSONObjectWithData:data options:0 error:nil];
        [[SocketIOManager sharedSocketIOManager] subscribes:socketDomain namspace:socketNameSpace subscribes:map];
    } else if ([SOCKET_UNSUBSCRIBES isEqualToString:call.method]) {
        NSString *socketData = call.arguments[SOCKET_DATA];
        NSData *data = [socketData dataUsingEncoding:NSUTF8StringEncoding];
        id map = [NSJSONSerialization JSONObjectWithData:data options:0 error:nil];
        [[SocketIOManager sharedSocketIOManager] unSubscribes:socketDomain namspace:socketNameSpace subscribes:map];
    } else if ([SOCKET_UNSUBSCRIBES_ALL isEqualToString:call.method]) {
        [[SocketIOManager sharedSocketIOManager] unSubscribesAll:socketDomain namspace:socketNameSpace];
    } else if ([SOCKET_SEND_MESSAGE isEqualToString:call.method]) {
        NSLog(@"SOCKET_SEND_MESSAGE native");
        NSString *event = call.arguments[SOCKET_EVENT];
        NSString *message = call.arguments[SOCKET_MESSAGE];
        if (event && message) {
            [[SocketIOManager sharedSocketIOManager] sendMessage:event message:message domain:socketDomain namspace:socketNameSpace callBackStatus:callback];
        }
    } else if ([SOCKET_DESTROY isEqualToString:call.method]) {
        [[SocketIOManager sharedSocketIOManager] destroySocketDomain:socketDomain namspace:socketNameSpace];
    } else if ([SOCKET_DESTROY_ALL isEqualToString:call.method]) {
        [[SocketIOManager sharedSocketIOManager] destroyAllSocket];
    } else {
        result(FlutterMethodNotImplemented);
    }
}

@end
